### PR TITLE
[#76] Protocol v3 & v4 support, serialization/deserialization rewrite using scodec

### PR DESCRIPTION
**Pull request by tolbertam**
For #76, Adds protocol v3 and v4 support.  This is a work in progress, but it's functional and currently all tests pass.

Summary of the major changes:
- Use [scodec](https://github.com/scodec/scodec) for serialization/deserialization.  All serde has been moved into a standalone `codec` module with scodec being the only dependency.
- Change Prime store implementations to keep the request intact instead of deconstructing on input and reconstructing when receiving via the route APIs.  This should keep things simpler.  Added a memoized `prime` member on the `Then` implementations to convert the inputs to a `Prime`.

There should be no functional changes for users with one slight exception.  Previously if no Prime is found for a query, query parameters would not be recorded in the activity log for batch and prepared statements.  Now it records the query parameters as if they were varchars.   There may also be some small nuances in conversions from json strings/numbers/booleans/etc. into the native type for a given data type, but i haven't identified any yet.

What is remaining:
- [x] Update [cql-antlr](https://github.com/scassandra/cql-antlr) to include the new types added in v3/v4 (tuples, udt, time, date, smallint, tinyint, etc.).  The codecs exist for these, but need to be tested.  I think there might be some barriers to supporting UDTs (covered in #159), so depending on the difficulty level I may defer this. (scassandra/cql-antlr#4)
- [x] Clean up, document and add more tests for the codec module.   scodec is really nice, but its API is intimidating from my view.   It is also not a complete implementation of the protocol, but the existing implementation isn't either (can maybe do more as follow ons).
- [x] Replace ServerStubrRunnerIntegrationTest / PrimingJsonHelperTest.  I removed these earlier, need to replace them with new alternatives.
- [x] Add/update headers (pending #170 merge)
- [x] Clean up commit history

### CODE REVIEW

**Comment by chbatey** [comment]
How's this going? 

Are you at the Cassandra summit? Might be nice to review together

**Comment by tolbertam** [comment]
Hi @chbatey.  It's going pretty well!   Existing tests are passing, what's left is adding additional tests in the codec module, adding additional tests for the new types and maybe some clean up.  Hopefully I can complete that really soon.

I'll be at the summit, I'd definitely be up for a review, that'd be fun!

**Comment by tolbertam** [comment]
Rebased against master and added license headers.  Working on finishing up tests in the codec module and updating rest of the code to support the v3/v4 types.  I'll probably get that all going and clean up the commit history today.

**Comment by chbatey** [comment]
Busy man!

**Comment by tolbertam** [comment]
rebased on master to get the `cql-antlr` changes.  I'll proceed to wire in the new types, and from there that should leave just some testing/doc updates/cleanup!

**Comment by tolbertam** [comment]
Getting closer, completed adding the new types and failures.  Going to spend some time tomorrow evening updating the docs.  I'll then do some cleanup / optimization of the codec code and clean up the commit history.  If all goes well should complete sometime this weekend.

**Comment by tolbertam** [comment]
I've spent some time the last few days adding integration tests for the new types and errors, and some more tests and documentation for the codec module.

I think i've got everything worked out and this is ready aside from the commit history.  @chbatey would you prefer I squash the commits now or would you have me leave it as is for now for review?

**Comment by mgobec** [comment]
Great work @tolbertam 👍

**Comment by chbatey** [comment]
Taking a look this weekend

File: server/src/main/scala/org/scassandra/server/actors/NativeProtocolMessageHandler.scala [hunk 1]
```diff
@@ -15,85 +15,62 @@
...
+        case Options =>
+          log.debug("Received options message. Sending to OptionsHandler")
+          optionsHandler forward message
+        case Startup(_) =>
+          val queryHandler = queryHandlerFactory(context)
+          val batchHandler = batchHandlerFactory(context, prepareHandler)
+          val registerHandler = registerHandlerFactory(context)
+          write(Ready, frame.header)
+          context become initialized(queryHandler, batchHandler, registerHandler)
+        case _ => {
```
Comment (chbatey) [inline server/src/main/scala/org/scassandra/server/actors/NativeProtocolMessageHandler.scala]: remove braces to make it consistent with the other cases

File: server/src/main/scala/org/scassandra/server/actors/ExecuteHandler.scala [hunk 2]
```diff
@@ -15,115 +15,67 @@
...
-  }
-
-  private def createMessage(preparedPrime: Prime, executeRequest: ExecuteRequest ,stream: Byte, msgFactory: CqlMessageFactory, connection: ActorRef) = {
-    preparedPrime.result match {
-      case SuccessResult => msgFactory.createRowsMessage(preparedPrime, stream)
-      case result: ErrorResult => msgFactory.createErrorMessage(result, stream, executeRequest.consistency)
-      case result: FatalResult => result.produceFatalError(connection)
+        writePrime(execute, prime, header, Some(connection), alternative=Some(Reply(VoidResult)), consistency = Some(execute.parameters.consistency))
+      case None =>
+        var errMsg = s"Could not find prepared statement with id: 0x${execute.id.toHex}"
```
Comment (chbatey) [inline server/src/main/scala/org/scassandra/server/actors/ExecuteHandler.scala]: should be a val

File: server/src/main/scala/org/scassandra/server/actors/ExecuteHandler.scala [hunk 3]
```diff
@@ -15,115 +15,67 @@
...
+        activityLog.recordPreparedStatementExecution(errMsg, execute.parameters.consistency, Nil, Nil)
```
Comment (chbatey) [inline server/src/main/scala/org/scassandra/server/actors/ExecuteHandler.scala]: i take it you found this useful for testing? just a little off we use an error msg as the statement text

**Comment by chbatey** [comment]
Happy to merge this whenever. I'll be adding some features shortly

File: server/src/main/scala/org/scassandra/server/actors/ExecuteHandler.scala (hunk 2 above)
Comment (tolbertam) [inline server/src/main/scala/org/scassandra/server/actors/ExecuteHandler.scala]: d'oh!  I've been doing too much javascript lately :smile:

File: server/src/main/scala/org/scassandra/server/actors/ExecuteHandler.scala (hunk 3 above)
Comment (tolbertam) [inline server/src/main/scala/org/scassandra/server/actors/ExecuteHandler.scala]: Hrmm, I may be misunderstanding, isn't the behavior the same as before?  I think the result is in this case that we record a prepared statement execution with the error message as the query text, and then an Unprepared message gets sent back to the client which will cause most drivers to repreare the statement.

**Comment by tolbertam** [comment]
added commit for review feedback, thanks!  Would you like me to do anything with the commit history, or would you just prefer to squash the whole thing on merge?

**Comment by chbatey** [comment]
Squashing into one if find!

It has been a long time since I looked at Scala lol

**Comment by tolbertam** [comment]
Just remembered, when it comes time to do a release with these changes, there are a few activites:

* Merge https://github.com/scassandra/cql-antlr/pull/7 to indicate that repository is deprecated in favor of cql-antlr being in this repository now.
* Update gh-pages branch since i made some doc updates.

**Comment by chbatey** [comment]
Thanks @tolbertam

**Comment by tolbertam** [comment]
@chbatey would you like me to merge and go through the process to release this?  Can do it tomorrow if you like.

**Comment by chbatey** [comment]
Go for it!

**Comment by tolbertam** [comment]
Awesome, thanks!  Will take care of that this evening :)

**Pull request merged**